File: backend/services/database/backup.js
/**
 * Database Backup Service
 * Handles automated backups and recovery for MongoDB
 */

const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const cron = require('node-cron');
const logger = require('../../config/logger');
const { getConfig } = require('../../config/env');

// Ensure backup directory exists
const initBackupDir = () => {
  const config = getConfig();
  const backupDir = config.backup.storagePath;

  if (!fs.existsSync(backupDir)) {
    fs.mkdirSync(backupDir, { recursive: true });
    logger.info(`Created backup directory at ${backupDir}`);
  }

  return backupDir;
};

// Format date for filename
const getFormattedDate = () => {
  const now = new Date();
  return `${now.getFullYear()}-${(now.getMonth() + 1)
    .toString()
    .padStart(2, '0')}-${now.getDate().toString().padStart(2, '0')}_${now
    .getHours()
    .toString()
    .padStart(2, '0')}-${now.getMinutes().toString().padStart(2, '0')}`;
};

// Perform database backup using mongodump
const performBackup = async () => {
  const config = getConfig();
  const backupDir = initBackupDir();
  const dateStr = getFormattedDate();
  const backupPath = path.join(backupDir, `backup-${dateStr}`);
  const mongoUri = config.mongodb.uri;

  if (!mongoUri) {
    logger.error('Cannot perform backup: MongoDB URI not configured');
    return null;
  }

  return new Promise((resolve, reject) => {
    // Create mongodump command
    // For production use, mongodump should be installed on the server
    const cmd = `mongodump --uri="${mongoUri}" --out="${backupPath}" --gzip`;

    logger.info(`Starting database backup to ${backupPath}`);
    
    exec(cmd, (error, stdout, stderr) => {
      if (error) {
        logger.error(`Backup failed: ${error.message}`);
        return reject(error);
      }
      
      if (stderr) {
        logger.warn(`Backup warnings: ${stderr}`);
      }
      
      logger.info(`Database backup completed successfully to ${backupPath}`);
      
      // After successful backup, clean up old backups
      cleanupOldBackups();
      
      resolve(backupPath);
    });
  });
};

// Clean up old backups to save space
const cleanupOldBackups = () => {
  const config = getConfig();
  const backupDir = config.backup.storagePath;
  const retentionDays = config.backup.retentionDays;
  
  if (!fs.existsSync(backupDir)) {
    return;
  }
  
  // Calculate cutoff date
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - retentionDays);
  
  fs.readdir(backupDir, (err, files) => {
    if (err) {
      logger.error(`Error reading backup directory: ${err.message}`);
      return;
    }
    
    files.forEach(file => {
      const filePath = path.join(backupDir, file);
      
      // Skip if not a directory
      if (!fs.statSync(filePath).isDirectory()) {
        return;
      }
      
      // Check if backup is older than retention period
      const stats = fs.statSync(filePath);
      const fileDate = new Date(stats.mtime);
      
      if (fileDate < cutoffDate) {
        // Remove old backup
        fs.rm(filePath, { recursive: true }, (err) => {
          if (err) {
            logger.error(`Error removing old backup ${filePath}: ${err.message}`);
          } else {
            logger.info(`Removed old backup: ${filePath}`);
          }
        });
      }
    });
  });
};

// Restore database from backup
const restoreFromBackup = async (backupPath) => {
  const config = getConfig();
  const mongoUri = config.mongodb.uri;
  
  if (!mongoUri) {
    logger.error('Cannot perform restore: MongoDB URI not configured');
    return false;
  }
  
  if (!backupPath || !fs.existsSync(backupPath)) {
    logger.error(`Backup path not found: ${backupPath}`);
    return false;
  }
  
  return new Promise((resolve, reject) => {
    // Create mongorestore command
    const cmd = `mongorestore --uri="${mongoUri}" "${backupPath}" --drop --gzip`;
    
    logger.info(`Starting database restore from ${backupPath}`);
    
    exec(cmd, (error, stdout, stderr) => {
      if (error) {
        logger.error(`Restore failed: ${error.message}`);
        return reject(error);
      }
      
      if (stderr) {
        logger.warn(`Restore warnings: ${stderr}`);
      }
      
      logger.info('Database restore completed successfully');
      resolve(true);
    });
  });
};

// Schedule automated backups
const scheduleBackups = () => {
  const config = getConfig();
  const cronSchedule = config.backup.schedule;
  
  // Skip scheduling in test environment
  if (process.env.NODE_ENV === 'test') {
    return false;
  }
  
  // Validate cron expression
  if (!cron.validate(cronSchedule)) {
    logger.error(`Invalid backup cron schedule: ${cronSchedule}`);
    return false;
  }
  
  logger.info(`Scheduling automated backups with schedule: ${cronSchedule}`);
  
  // Schedule backup job
  cron.schedule(cronSchedule, async () => {
    try {
      await performBackup();
    } catch (error) {
      logger.error(`Scheduled backup failed: ${error.message}`);
    }
  });
  
  return true;
};

// List available backups
const listBackups = () => {
  const config = getConfig();
  const backupDir = config.backup.storagePath;
  
  if (!fs.existsSync(backupDir)) {
    return [];
  }
  
  try {
    const files = fs.readdirSync(backupDir);
    
    // Filter directories only and sort by date (newest first)
    const backups = files
      .filter(file => {
        const filePath = path.join(backupDir, file);
        return fs.statSync(filePath).isDirectory();
      })
      .map(file => {
        const filePath = path.join(backupDir, file);
        const stats = fs.statSync(filePath);
        return {
          name: file,
          path: filePath,
          date: stats.mtime,
        };
      })
      .sort((a, b) => b.date - a.date);
    
    return backups;
  } catch (error) {
    logger.error(`Error listing backups: ${error.message}`);
    return [];
  }
};

module.exports = {
  performBackup,
  restoreFromBackup,
  scheduleBackups,
  listBackups,
  cleanupOldBackups,
  initBackupDir
}; 

File: backend/app/utils/backup.py
import os
import shutil
import time
import glob
import asyncio
import subprocess
from datetime import datetime, timedelta
from pathlib import Path

from app.utils.logging import get_logger
from app.config.settings import get_settings

settings = get_settings()
logger = get_logger("database.backup")

class DatabaseBackup:
    """Handle MongoDB database backup operations"""
    
    def __init__(self):
        self.backup_dir = Path(settings.backup_storage_path)
        self.db_name = settings.mongodb_db_name
        self.mongodb_uri = settings.mongodb_uri
        self.retention_days = settings.backup_retention_days
        
        # Create backup directory if it doesn't exist
        self.backup_dir.mkdir(exist_ok=True, parents=True)
    
    async def create_backup(self) -> bool:
        """Create a database backup using mongodump"""
        try:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            backup_path = self.backup_dir / f"backup_{self.db_name}_{timestamp}"
            
            # Ensure the directory exists
            backup_path.mkdir(exist_ok=True)
            
            # Build mongodump command
            cmd = [
                "mongodump",
                f"--uri={self.mongodb_uri}",
                f"--db={self.db_name}",
                f"--out={backup_path}"
            ]
            
            # Execute mongodump
            logger.info(f"Starting backup of {self.db_name} database")
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            stdout, stderr = await process.communicate()
            
            if process.returncode == 0:
                # Compress the backup
                archive_name = f"{backup_path}.tar.gz"
                shutil.make_archive(
                    str(backup_path),
                    'gztar',
                    root_dir=str(self.backup_dir),
                    base_dir=backup_path.name
                )
                
                # Remove the uncompressed directory
                shutil.rmtree(backup_path)
                
                logger.info(f"Successfully created backup: {archive_name}")
                return True
            else:
                error_msg = stderr.decode() if stderr else "Unknown error"
                logger.error(f"Backup failed: {error_msg}")
                return False
                
        except Exception as e:
            logger.error(f"Error creating backup: {str(e)}", exc_info=True)
            return False
    
    async def restore_backup(self, backup_file: str) -> bool:
        """Restore a database from backup"""
        try:
            # Verify the backup file exists
            backup_path = self.backup_dir / backup_file
            if not backup_path.exists():
                logger.error(f"Backup file not found: {backup_file}")
                return False
            
            # Extract the backup if it's compressed
            if backup_file.endswith('.tar.gz'):
                extract_path = self.backup_dir / f"restore_temp_{int(time.time())}"
                extract_path.mkdir(exist_ok=True)
                
                shutil.unpack_archive(
                    str(backup_path),
                    extract_dir=str(extract_path)
                )
                
                # Find the extracted directory
                extracted_dirs = [d for d in extract_path.iterdir() if d.is_dir()]
                if not extracted_dirs:
                    logger.error("No valid backup found in archive")
                    return False
                
                restore_path = extracted_dirs[0]
            else:
                restore_path = backup_path
            
            # Build mongorestore command
            cmd = [
                "mongorestore",
                f"--uri={self.mongodb_uri}",
                "--drop",  # Drop existing collections before restoring
                str(restore_path)
            ]
            
            # Execute mongorestore
            logger.info(f"Starting restore of {self.db_name} database from {backup_file}")
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            stdout, stderr = await process.communicate()
            
            # Clean up extracted files if needed
            if backup_file.endswith('.tar.gz'):
                shutil.rmtree(extract_path)
            
            if process.returncode == 0:
                logger.info(f"Successfully restored database from {backup_file}")
                return True
            else:
                error_msg = stderr.decode() if stderr else "Unknown error"
                logger.error(f"Restore failed: {error_msg}")
                return False
                
        except Exception as e:
            logger.error(f"Error restoring backup: {str(e)}", exc_info=True)
            return False
    
    async def cleanup_old_backups(self) -> int:
        """Delete backups older than retention period"""
        try:
            # Calculate cutoff date
            cutoff_date = datetime.now() - timedelta(days=self.retention_days)
            removed_count = 0
            
            # Find all backup files
            backup_files = list(self.backup_dir.glob("backup_*.tar.gz"))
            
            for backup_file in backup_files:
                try:
                    # Extract date from filename (format: backup_dbname_YYYYMMDD_HHMMSS.tar.gz)
                    date_str = backup_file.stem.split('_')[2:4]
                    if len(date_str) >= 2:
                        date_str = '_'.join(date_str)
                        backup_date = datetime.strptime(date_str, "%Y%m%d_%H%M%S")
                        
                        # Check if backup is older than retention period
                        if backup_date < cutoff_date:
                            backup_file.unlink()
                            removed_count += 1
                            logger.info(f"Removed old backup: {backup_file.name}")
                except (ValueError, IndexError):
                    # Skip files that don't match expected naming pattern
                    logger.warning(f"Could not parse date from backup filename: {backup_file.name}")
                    continue
            
            logger.info(f"Cleanup complete. Removed {removed_count} old backups.")
            return removed_count
            
        except Exception as e:
            logger.error(f"Error cleaning up old backups: {str(e)}", exc_info=True)
            return 0
    
    async def list_available_backups(self):
        """List all available backups sorted by date (newest first)"""
        try:
            backup_files = list(self.backup_dir.glob("backup_*.tar.gz"))
            
            # Sort backups by creation time (newest first)
            backup_files.sort(key=lambda x: x.stat().st_ctime, reverse=True)
            
            # Create backup info list
            backups = []
            for backup_file in backup_files:
                size_mb = backup_file.stat().st_size / (1024 * 1024)
                created = datetime.fromtimestamp(backup_file.stat().st_ctime)
                
                backups.append({
                    "filename": backup_file.name,
                    "size_mb": round(size_mb, 2),
                    "created_at": created.isoformat(),
                    "age_days": (datetime.now() - created).days
                })
            
            return backups
            
        except Exception as e:
            logger.error(f"Error listing backups: {str(e)}", exc_info=True)
            return [] 